pkg/query/callgraph.go

```diff
@@ -165,3 +170,114 @@ func lineToGraphNode(
 		},
 	}
 }
+
+func prunableNodes(nodes []*querypb.CallgraphNode, c int64) []*querypb.CallgraphNode {
+	if len(nodes) == 0 {
+		return nodes
+	}
+	sort.Slice(nodes, func(i, j int) bool {
+		return nodes[i].Cumulative > nodes[j].Cumulative
+	})
+	i := 0
+	cutoffValue := (float64(c) * NodeCutOffFraction)
+	for ; i < len(nodes); i++ {
+		if float64(nodes[i].Cumulative) < cutoffValue {
+			break
+		}
+	}
+	return nodes[i:]
+}
+
+func pruneGraph(graph *querypb.Callgraph) *querypb.Callgraph {
+	prunableNodes := prunableNodes(graph.Nodes, graph.Cumulative)
+	finalNodes := make([]*querypb.CallgraphNode, 0)
+	finalEdges := make([]*querypb.CallgraphEdge, 0)
+	edgesMap := make(map[string]*querypb.CallgraphEdge)
+	incomingEdges := make(map[string][]*querypb.CallgraphEdge)
+	outgoingEdges := make(map[string][]*querypb.CallgraphEdge)
+	prunableNodesMap := make(map[string]bool)
+	for _, edge := range graph.Edges {
+		if incomingEdges[edge.Target] == nil {
+			incomingEdges[edge.Target] = []*querypb.CallgraphEdge{}
+		}
+		incomingEdges[edge.Target] = append(incomingEdges[edge.Target], edge)
+
+		if outgoingEdges[edge.Source] == nil {
+			outgoingEdges[edge.Source] = []*querypb.CallgraphEdge{}
+		}
+		outgoingEdges[edge.Source] = append(outgoingEdges[edge.Source], edge)
+		edgesMap[edge.Id] = edge
+	}
+
+	for _, node := range prunableNodes {
+		prunableNodesMap[node.Id] = true
+	}
+
+	nodesToRemove := make(map[string]bool, 0)
+	edgesToRemove := make(map[string]bool, 0)
+	edgesToCreate := make([]*querypb.CallgraphEdge, 0)
+
+	// Validate the eligibility of each prunableNode
+	for _, node := range prunableNodes {
+		if !prunableNodesMap[node.Id] {
+			continue
+		}
+		if len(incomingEdges[node.Id]) > 1 {
+			// Cannot prune nodes with multiple incoming edges.
+			continue
+		}
+		if len(incomingEdges[node.Id]) == 0 || len(outgoingEdges[node.Id]) == 0 {
+			// Cannot prune leaf nodes.
+			continue
+		}
+		nodesToRemove[node.Id] = true
+	}
+
+	// Remove nodes and identify edges to patch
+	for _, node := range graph.Nodes {
+		if nodesToRemove[node.Id] {
+			// patch the edges from its parent to child nodes
+			parentNodeId, cummValue, incomingEdgesToRemove := findAValidParent(node, incomingEdges, outgoingEdges, nodesToRemove)
+			for _, edge := range outgoingEdges[node.Id] {
+				if nodesToRemove[edge.Target] {
+					// Skipping the edge creation as this will be patched by the downstream node that is being removed.
+					continue
+				}
+				newEdge := &querypb.CallgraphEdge{Id: uuid.New().String(), Source: parentNodeId, Target: edge.Target, Cumulative: cummValue}
```

Comment:
In pprof the edge seems to know when nodes were skipped and draws a dotted line (I think this is useful). To do that I think we need some form of indicator on the edge that this happened, no?

Reply:
Yes, that's a good point. Will add a field `isCollapsed` and set it to `true` on these edges?

Reply:
Done, implemented it.